### ISSUE TITLE
Large gitignore slows down operations a lot

### ISSUE THREAD

**Issue by lacostej** [git-lfs/git-lfs#3668]
The way git LFS manages its filtering mechanisms in relation to the gitignore file slows down operations a lot.

E.g. an amend to a commit touching a single small file takes 4 seconds with a 800 lines `.gitignore` and 6 seconds with a 1200 lines file.

Our repo has 30k files, of which 5k are LFS + 7k ignored files and a large gitignore file (~1000 lines).

Tracing reveals that most of the time is spent rejecting / accepting files

```
$ GIT_TRACE=1 git commit -m "XXX" --amend 2>&1 | tee git.log
$ grep accepting git.log | head -1
08:51:25.676481 trace git-lfs: filepathfilter: accepting "AAAA"
$ grep accepting git.log | tail -1
08:51:31.910054 trace git-lfs: filepathfilter: accepting "BBBB"
```

(6 seconds between the first and last filepathfilter operations)

Some reading of the code makes me think there's a sub-optimal implementation of `Walk()` with regard to large gitignores.

https://github.com/git-lfs/git-lfs/blob/06a0c388056b615f165352dcdd27401a35f0a74f/tools/filetools.go#L338-L401

**Comment by bk2204** [comment]
Hey, thanks for writing in.

I expect there are definitely some optimizations we could make in this area. Git, for example, has optimizations to check if the item is a directory and ignore items which are not. I think there's been some work on improving some of our correctness which may help us improve performance here as well, by enabling some of the same optimizations that Git has.

I'll flag this as an enhancement and see if I can get around to looking at it.

**Comment by jerick1976** [comment]
Just wanted to give this issue a bit of a bump.  We have a number of large Git repositories with thousands of git-lfs objects that are slowing down on push due to the pre-push scan operations that git-lfs is doing.  One of the worst is a repository with 20k git-lfs objects that is using .gitignore files.  That repo recently required one hour to push three new binaries that were relatively small.

**Comment by bk2204** [comment]
I believe this has been fixed by #3823.
